lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -333,21 +333,31 @@ static LogicalResult rewriteWarpOpToScfFor(
     Type resultType = warpOp->getResultTypes()[it.index()];
     Value buffer = allocBuffer(val.getType());
 
-    // Store yielded vector into buffer.
+    // Store yielded value into buffer.
     rewriter.setInsertionPoint(yieldOp);
-    rewriter.create<vector::StoreOp>(yieldLoc, val, buffer, c0);
+    if (val.getType().isa<VectorType>()) {
+      rewriter.create<vector::StoreOp>(yieldLoc, val, buffer, c0);
+    } else {
+      rewriter.create<memref::StoreOp>(yieldLoc, val, buffer, c0);
+    }
 
-    // Load vector from buffer (after warpOp).
+    // Load value from buffer (after warpOp).
     rewriter.setInsertionPointAfter(ifOp);
-    auto loadedVectorType = resultType.cast<VectorType>();
-    int64_t loadSize = loadedVectorType.getShape()[0];
-
-    // loadOffset = laneid * loadSize
-    Value loadOffset = rewriter.create<arith::MulIOp>(loc, warpOp.laneid(),
-        rewriter.create<arith::ConstantIndexOp>(loc, loadSize));
-    Value loadOp = rewriter.create<vector::LoadOp>(loc, loadedVectorType,
-                                                   buffer, loadOffset);
-    replacements.push_back(loadOp);
+    if (resultType == val.getType()) {
+      // Result type and yielded value type are the same. This is a broadcast.
+      Value loadOp = rewriter.create<memref::LoadOp>(loc, buffer, c0);
+      replacements.push_back(loadOp);
+    } else {
+      auto loadedVectorType = resultType.cast<VectorType>();
+      int64_t loadSize = loadedVectorType.getShape()[0];
+
+      // loadOffset = laneid * loadSize
+      Value loadOffset = rewriter.create<arith::MulIOp>(loc, warpOp.laneid(),
+          rewriter.create<arith::ConstantIndexOp>(loc, loadSize));
+      Value loadOp = rewriter.create<vector::LoadOp>(loc, loadedVectorType,
+                                                     buffer, loadOffset);
+      replacements.push_back(loadOp);
```

Comment:
Looks good, eventually we should base all this on affine maps and that should generalize things, but at this point it is probably premature and this looks simpler that way.